sprint-1/step-2

ConfidentPie opened the pull request.

ConfidentPie committed:
Скопированные файлы

ArslanMustafin commented on src/components/tabs/tabs.jsx:
```diff
@@ -0,0 +1,21 @@
+function Tabs() {
+  const [current, setCurrent] = React.useState('buns')
+  return (
+    <div style={{ display: 'flex', marginBottom: '40px' }}>
```
**Надо исправить:**

Не стоит писать стили инлайново, в будущем это может затруднить поиск ошибок, и перебить стили заданных классов. Лучше дать какой-то класс, и там указать уже нужные вам стили.

ArslanMustafin commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,45 @@
+import BurgerIngredients from '../burger-ingredients/burger-ingredients';
+import BurgerConstructor from '../burger-constructor/burger-constructor';
+
+const API_ADDRESS = 'https://norma.nomoreparties.space/api/ingredients';
```
**Надо исправить:**

вынесите базовый адрес сервера `https://norma.nomoreparties.space/api` сервера в константу в отдельный файл `utils/burger-api.js`, а адрес эндпоинта можно оставить в самой функции запроса `${BURGER_API_URL}/ingredients`

ArslanMustafin commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,45 @@
+  useEffect(() => {
+    const getProductData = async () => {
+      try {
+        const res = await fetch(API_ADDRESS);
+
+        if (!res.ok) {
+          throw new Error(
+            `Зпрос не выполнен. Ошибка при выполнении запроса: ${res.status}`
+          );
+        }
+
+        const data = await res.json();
```
**Надо исправить:**

в дальнейшем добавятся ещё запросы к бэкенду и лучше вынести описание этих запросов в функции

Например:

```js
function getIngredients() {
   return fetch(`${NORMA_API}/ingredients`)
    .then(проверка ответа)
}
```

и разместить их в отдельном файле `utils/burger-api.js`, а не описывать запросы в самом компоненте

Т.к. проверка нужна у всех запросов, лучше вынести её в отдельную функцию

```js
const checkResponse = res => {
  return res.ok ? res.json() : res.json().then(err => Promise.reject(err));
};
```

`fetch(someUrl).then(checkResponse)`

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+} from '@ya.praktikum/react-developer-burger-ui-components';
+import BurgerElements from '../burger-elements/burger-elements';
+
+BurgerConstructor.propTypes = {
+  _id: PropTypes.string,
+  name: PropTypes.string,
+  type: PropTypes.string,
+  proteins: PropTypes.number,
+  fat: PropTypes.number,
+  carbohydrates: PropTypes.number,
+  calories: PropTypes.number,
+  price: PropTypes.number,
+  image: PropTypes.string,
+  image_mobile: PropTypes.string,
+  image_large: PropTypes.string,
+  __v: PropTypes.number,
+};
```
**Надо исправить:**

Во многих компонентах будет дублироваться описание PropTypes ингредиента.
Следует вынести его в отдельный файл (можно создать файл `prop-types.js` в папке `utils`) и переиспользовать описание пропсов ингредиента, а не дублировать;

а также у всех полей должно быть `isRequired`.

и сразу насчёт пропсов для массива ингредиентов. Необходимо указывать .isRequired как у элемента массива, так и у самого массива в propTypes.

Пример как правильно:

```js
data: PropTypes.arrayOf(ingredientType.isRequired).isRequired;
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+  return (
+    <>
+      <section className={`${burgerConstructor.container} pt-25 pl-4 pr-4`}>
+        <BurgerTop />
+        <ul className={`${burgerConstructor.list} custom-scroll`}>
+          <BurgerElements />
+        </ul>
+        <BurgerBottom />
```
**Надо исправить:**

не стоит выносить булки и списко в отдельные компоненты, там нету какой-то сложной логики.

Лучше написать так: 

```js
  <div className={`${styles.element} mb-4 mr-4`}>
    <ConstructorElement
      type="top"
      isLocked={true}
      text={`${bun.name} (верх)`}
      price={bun.price}
      thumbnail={bun.image}
    />
  </div>
  <ul className={styles.elements}>
    {ingredients.map((item, index) => {
      return (
        <li className={`${styles.element} mb-4 mr-2`} key={index}>
          <DragIcon />
          <div className={`${styles.element_fullwidth} ml-2`}>
            <ConstructorElement
              text={item.name}
              price={item.price}
              thumbnail={item.image}
            />
          </div>
        </li>
      );
    })}
  </ul>
  <div className={`${styles.element} mt-4 mr-4`}>
    <ConstructorElement
      type="bottom"
      isLocked={true}
      text={`${bun.name} (низ)`}
      price={bun.price}
      thumbnail={bun.image}
    />
  </div>
```

А `bun` и `ingredients` можно получать так.

```js
const { bun, ingredients } = useMemo(() => {
  return {
    bun: data.find(item => item.type === 'bun'),
    ingredients: data.filter(item => item.type !== 'bun'),
  };
}, [data]);
```

Т.к. вы сделали запросы к серверу, то нужно убрать захардкоженные данные и использовать данные с сервера.

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+      </section>
+      {isModalOpen && (
+        <Modal onClose={closeModal}>
+          <OrderDetails onClose={closeModal} />
```
**Надо исправить:**

Зачем тут `onClose={closeModal}`? не совсем понимаю.

ArslanMustafin commented on src/components/burger-elements/burger-bottom.jsx:
```diff
@@ -0,0 +1,19 @@
+const BurgerBottom = () => {
+  return (
+    <div className={burgersElement.top}>
+      <ConstructorElement
+        type='bottom'
+        isLocked={true}
+        text='Краторная булка N-200i (низ)'
+        price={200}
+        thumbnail={'https://code.s3.yandex.net/react/code/bun-02-mobile.png'}
+      />
```
**Надо исправить:**

Т.к. вы сделали запросы к серверу, то нужно убрать захардкоженные данные и использовать данные с сервера.

ArslanMustafin commented on src/components/burger-elements/burger-elements.jsx:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <>
+     {choosenIngridients.map(({ name, price, _id, image_mobile, type }) => {
+        if (type === "sauce" || type === "main") {
```
**Можно лучше:**

показал в `burger-constructor` как можно сделать лучше.

ArslanMustafin commented on src/components/burger-elements/burger-elements.jsx:
```diff
@@ -0,0 +1,36 @@
+  ConstructorElement,
+  DragIcon
+} from '@ya.praktikum/react-developer-burger-ui-components';
+import data from "../../utils/data"
```
**Надо исправить:**

Т.к. вы сделали запросы к серверу, то нужно убрать захардкоженные данные и использовать данные с сервера.

ArslanMustafin commented on src/components/burger-elements/burger-top.jsx:
```diff
@@ -0,0 +1,20 @@
+const BurgerTop = () => {
+    return (
+        <div className={burgersElement.top}>
+        <ConstructorElement
+          type="top"
+          isLocked={true}
+          text="Краторная булка N-200i (верх)"
+          price={200}
+          thumbnail={'https://code.s3.yandex.net/react/code/bun-02-mobile.png'}
+        />
```
**Надо исправить:**

Т.к. вы сделали запросы к серверу, то нужно убрать захардкоженные данные и использовать данные с сервера.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,64 @@
+import Modal from '../modal/modal';
+import IngredientDetails from '../ingredient-details/ingredient-details';
+
+BurgerIngredients.propTypes = {
+  _id: PropTypes.string,
+  name: PropTypes.string,
+  type: PropTypes.string,
+  proteins: PropTypes.number,
+  fat: PropTypes.number,
+  carbohydrates: PropTypes.number,
+  calories: PropTypes.number,
+  price: PropTypes.number,
+  image: PropTypes.string,
+  image_mobile: PropTypes.string,
+  image_large: PropTypes.string,
+  __v: PropTypes.number,
+};
```
**Надо исправить:**

Неправильно описаны пропсы.

Аналогичное замечание есть в `BurgerConstructor`

ArslanMustafin commented on src/components/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,30 @@
+  Counter
+} from '@ya.praktikum/react-developer-burger-ui-components';
+
+const IngredientCard = ({ ingredient, onCardClick }) => {
```
**Надо исправить:**

Для этого компонента так же нужно описать PropTypes.

ArslanMustafin commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import ingredientDetails from './ingredient-details.module.css';
+
+function IngredientDetails({ ingredient }) {
```
**Надо исправить:**

Для этого компонента так же нужно описать PropTypes.

ArslanMustafin commented on src/components/ingredients-list/ingredients-list.jsx:
```diff
@@ -0,0 +1,37 @@
+import ingredientsList from './ingredients-list.module.css';
+import IngredientCard from '../ingredient-card/ingredient-card';
+
+const IngredientsGroup = ({ title, ingredients, onCardClick }) => {
+
+  return (
+    <div className='mb-10'>
+      <h2 className={`${ingredientsList.title} mb-6`}>{title}</h2>
+      <ul className={ingredientsList.list}>
+        {ingredients.map(ingredient => (
+          <li key={ingredient._id}>
+            <IngredientCard ingredient={ingredient} onCardClick={onCardClick}/>
+          </li>
+        ))}
+      </ul>
+    </div>
+  );
+};
```
**Надо исправить:**

нужно вынести в отдельный компонент (папку, файл). А также описать PropTypes.

ArslanMustafin commented on src/components/ingredients-list/ingredients-list.jsx:
```diff
@@ -0,0 +1,37 @@
+};
+
+const IngredientList = ({ ingredients, onCardClick }) => {
+    const groupedIngredients = ingredients.reduce((acc, ingredient) => {
+      const { type } = ingredient;
+      acc[type] = [...(acc[type] || []), ingredient];
+      return acc;
+    }, {});
+
+    return (
+      <>
+        <IngredientsGroup title="Булки" ingredients={groupedIngredients.bun || []} onCardClick={onCardClick}/>
+        <IngredientsGroup title="Соусы" ingredients={groupedIngredients.sauce || []} onCardClick={onCardClick}/>
+        <IngredientsGroup title="Ингредиенты" ingredients={groupedIngredients.main || []} onCardClick={onCardClick}/>
+      </>
+    );
```
**Надо исправить:**

В данном случае лучше всего будет оставить это в `burger-ingredients`, а то у вас слишком много передач пропсов, что не очень хорошо)) 

А также оберните `groupedIngredients` в `useMemo`, когда изучите.

ArslanMustafin commented on src/components/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,12 @@
+import modalOverlay from './modal-overlay.module.css';
+
+const ModalOverlay = ({ onClose }) => {
+  const handleClick = () => {
+    onClose();
+  };
```
**Надо исправить:**

Кажется, эта функция избыточна.
Можно сразу написать `onClick={onClose}`

ArslanMustafin commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,42 @@
+      <ModalOverlay onClose={onClose} />
+      <div
+        className={modal.content}
+        onClick={(e) => e.stopPropagation()}
```
**Надо исправить:**

В данном случае тут не нужно дополнительно обрабатывать клик. Контент перекрывает `modalOverlay` (контент стоит выше), поэтому при клике по контенту, обработчик клика `modalOverlay` не будет работать.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа.

Приложение начинает обрастать функционалом и уже постепенно проглядывается картина готового продукта.

### **Вы молодец!**

- Декомпозиция файлов по директориям соответствует описанию;
- Сборка и запуск проекта выполняются без ошибок;
- Есть модальные окна;
- Данные получают с сервера;

### **Но есть несколько замечаний, которые нужно исправить:**

- Вам нужно добавить файл `.gitignore`.

- Нужно очистить проект от лишних файлов, которые остались при инициализации create-react-app.

  - App.test.tsx

- <https://disk.yandex.com/i/uQqJYKqScU7WNA>, активной должна быть только 1 ссылка, а не все.

- Не стоит писать стили инлайново, в будущем это может затруднить поиск ошибок, и перебить стили заданных классов. Лучше дать какой-то класс, и там указать уже нужные вам стили.

### **Сделает код лучше:**

- Необходимо указывать .isRequired как у элемента массива, так и у самого массива в propTypes.

  Пример как правильно:

  ```js
  data: PropTypes.arrayOf(ingredientTypes.isRequired).isRequired;
  ```

  Проверьте во всём проекте.

### **_Подробно эти и более мелкие замечания, отмечены в коде._**

#### **Хочу напомнить, что работа может быть принята, только после исправления всех "Надо исправить". Перед отправкой на ревью, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12).**

ConfidentPie committed:
First fix

1. Adding gitignore
2. Fixing inline styles in tabs.jsx
3. Create burger-api.js

ConfidentPie committed:
fix 2. Fixing and adding PropTypes

ConfidentPie committed:
Fix 3.

ConfidentPie committed:
fixing  burger-ingredient

ConfidentPie committed:
fixing burger-constructor

ConfidentPie committed:
deleting app.test.tsx

ArslanMustafin commented on src/components/burger-ingredient/burger-ingredient.jsx:
```diff
@@ -0,0 +1,24 @@
+} from '@ya.praktikum/react-developer-burger-ui-components';
+
+
+const BurgerElements = ({item}) => {
```
**Надо исправить:**

Для этого компонента так же нужно описать PropTypes.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,64 +1,70 @@
+      const { type } = ingredient;
+      acc[type] = [...(acc[type] || []), ingredient];
+      return acc;
+    }, {});
```
**Можно лучше:**

Думаю намного лучше будет сразу определить ключи со значением пустого массива. Тогда не нужно будет делать условие `|| []` в `jsx`.

ArslanMustafin commented on src/utils/prop-types.js:
```diff
@@ -0,0 +1,18 @@
+import PropTypes from 'prop-types';
+
+  export const ingredientsType = PropTypes.shape({
```
**Надо исправить:**

у всех полей нужно указывать `.isRequired`.

ArslanMustafin commented on src/components/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,35 @@
+  );
+};
+
+IngredientCard.propTypes = {
+  ingredient: ingredientsType.isRequired,
+};
```
**Надо исправить:**

Описаны не все пропсы.

`onCardClick`

ArslanMustafin commented on src/components/ingredients-group/ingredients-group.jsx:
```diff
@@ -0,0 +1,27 @@
+    );
+  };
+
+  IngredientsGroup.propTypes = {
+    ingredients: ingredientsArrayType,
+  };
```
**Надо исправить:**

описаны не все пропсы.

`title, onCardClick`

ArslanMustafin commented on src/components/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import modalOverlay from './modal-overlay.module.css';
+
+const ModalOverlay = ({ onClose }) => {
```
**Надо исправить:**

Для этого компонента так же нужно описать PropTypes.

ArslanMustafin commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,41 @@
+import ModalOverlay from '../modal-overlay/modal-overlay';
+import { CloseIcon } from '@ya.praktikum/react-developer-burger-ui-components';
+
+const Modal = ({ children, onClose }) => {
```
**Надо исправить:**

Для этого компонента так же нужно описать PropTypes.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа. 🔥

### **Вы молодец!**

- Практически все замечания были исправлены;

### **Но есть несколько замечаний, которые нужно исправить:**

- Отметил замечания в коде.

ConfidentPie committed:
PropTypes fixing

ConfidentPie committed:
Final fixing of PropsType

ArslanMustafin commented on src/components/burger-ingredient/burger-ingredient.jsx:
```diff
@@ -21,4 +23,8 @@ const BurgerElements = ({item}) => {
 
-export default BurgerElements;
+burgersIngredient.propTypes = {
+  item: PropTypes.object.isRequired
```
**Рекомендую исправить:**

У вас ведь есть тип ингредиента `ingredientType`.  Используйте его.

ArslanMustafin reviewed (approved):
## Code Review

Здравствуйте!

### **Вы молодец! Работа принята, удачи в дальнейшем обучении!**

- Все замечания были исправлены;
